io/disk/parallel_dd.py

```diff
@@ -0,0 +1,231 @@
+        :params dd_woptions: dd write options.
+        :params dd_roptions: dd read options.
+        :params fs_dd_woptions: dd write in streams.
+        :params fs_dd_roptions: dd read in streams.
+        """
+
+        self.disk = self.params.get('disk')
+        if not self.disk:
+            self.error('Test requires disk parameter,Please check README')
+        self.fsys = partition_lib.Partition(self.disk, mountpoint=self.srcdir)
+        self.megabytes = self.params.get('megabytes', default=100)
+        self.blocks = self.params.get('blocks', default=None)
+        self.streams = self.params.get('streams', default=2)
+        self.blocks_per_file = self.params.get(
+            'blocks_per_file', default=None)
+        self.fstype = self.params.get('fstype', default=None)
+        self.seq_read = self.params.get('seq_read', default=True)
+        self.dd_woptions = self.params.get('dd_woptions', default='')
+        self.dd_roptions = self.params.get('dd_roptions', default='')
+        self.fs_dd_woptions = self.params.get('fs_dd_woptions', default='')
+        self.fs_dd_roptions = self.params.get('fs_dd_roptions', default='')
+        if not self.blocks:
+            self.blocks = self.megabytes * 256
+
+        if not self.blocks_per_file:
+            self.blocks_per_file = self.blocks / self.streams
+
+        root_fs_device_cmd = "df | egrep /$ | awk '{print $1}'"
+        root_fs_device = process.system_output(root_fs_device_cmd, shell=True)
+        self.root_fstype = self._device_to_fstype('/etc/fstab', root_fs_device)
+
+        if not self.fstype and self.root_fstype:
+            self.fstype = self.root_fstype
+
+        self.old_fstype = self._device_to_fstype('/etc/mtab')
+        if not self.old_fstype:
+            self.old_fstpye = self._device_to_fstype('/etc/fstab')
+        if not self.old_fstype:
+            self.old_fstype = self.fstype
+        self.log.info('Dumping %d megabytes across %d streams', self.megabytes,
+                      self.streams)
+
+    def raw_write(self):
+        """
+        Performs raw write operation.
+        """
+        self.log.info("Timing raw write of %d megabytes", self.megabytes)
+        cmd = 'dd if=/dev/zero of=%s bs=4k count=%d' % (self.fsys.device,
+                                                        self.blocks)
+
+        for option in self.dd_woptions.split():
+            cmd += " %s=%s" % (option.split(":")[0], option.split(":")[1])
+        process.run(cmd, shell=True)
+
+    def raw_read(self):
+        """
+        Performs raw read operation.
+        """
+        self.log.info("Timing raw read of %d megabytes", self.megabytes)
+        cmd = 'dd if=%s of=/dev/null bs=4k count=%d' % (self.fsys.device,
+                                                        self.blocks)
+        for option in self.dd_roptions.split():
+            cmd += " %s=%s" % (option.split(":")[0], option.split(":")[1])
+        process.run(cmd + ' > /dev/null', shell=True)
+
```

Comment:
Can you do something like this, to overload the function for read and write..

    def raw_io(self, operation=''):
        """
        Performs raw io operation.
        """
        self.log.info("Timing raw %s of %d megabytes", % (operation, self.megabytes))
        cmd = 'dd '
        if operation == 'read':
            cmd += 'if=%s of=/dev/zero ' % self.fsys.device
            option = self.dd_roptions
        if operation == 'write':
            option = self.dd_woptions
            cmd += 'if=/dev/zero of=%s ' % self.fsys.device
        cmd += 'bs=4k count=%d' % self.blocks

        for option in options.split():
            cmd += " %s=%s" % (option.split(":")[0], option.split(":")[1])
        process.run(cmd, shell=True)

Note: untested code